File: 012/haskell/src/Problem12.hs
module Problem12 where

import Data.Map (Map)
import qualified Data.Map as Map
import Data.Maybe (fromMaybe)
import Data.Numbers.Primes (primes)

answer :: Integer
answer = head $ filter hasEnoughFactors $ triangles where
    hasEnoughFactors n = numberOfFactors n > 500

triangles :: [Integer]
triangles = scanl1 (+) [1..]

numberOfFactors :: Integer -> Integer
numberOfFactors n = product $ map (+1) $ Map.elems $ factor n

-- Maps prime factors to cardinalities
factor :: Integer -> Map Integer Integer
factor n = factor' Map.empty n where
    factor' f' 1 = f'
    factor' f' n' = factor' f'' n'' where
        f'' = Map.alter ((Just).(+1).(fromMaybe 0)) d f'
        n'' = (n' `div` d)
        d = lowestPrimeFactor n'

lowestPrimeFactor :: Integer -> Integer
lowestPrimeFactor n = head $ filter (`divides` n) primes where
    n' `divides` d = d `mod` n' == 0

File: 012/scala/src/main/scala/Problem12.scala
object Problem12 {

  println("hi")

  lazy val primes: Stream[BigInt] =
    Stream.iterate(BigInt(2))(_ + 1)
      .filter(_.isProbablePrime(45))

  lazy val triangles: Stream[BigInt] =
    Stream.iterate(BigInt(1))(_ + 1)
      .scanLeft(BigInt(0))(_ + _)

  def nrOfFactors(n: BigInt): Int = {
    val fs = new collection.mutable.HashMap[BigInt, Int]
    var i = n
    while (i != BigInt(1)) {
      val p = primes.find(p => i % p == 0).get
      fs.put(p, fs.getOrElse(p, 0) + 1)
      i /= p
    }
    fs.values.map(1 + _).product
  }

  lazy val answer: Int =
    triangles.drop(1).filter(nrOfFactors(_) > 500).head.toInt

}
